File: dedupfs/src/chunk.rs
use blake3;
use anyhow::Result;
use std::any::Any;
use std::collections::HashMap;
use std::sync::{RwLock, LazyLock};
use fastcdc::v2020::FastCDC;
use serde::{Deserialize, Serialize};
use tracing::{error, info};
use crate::config::ChunkConfig;
use crate::vfile::DedupFS;
use crate::kvstore::{make_prefixed_key, key_prefix};
use crate::block::read_block;
use crate::cache::{CacheItem, new_shared_cache, SharedCache};
use crate::errors::{ChunkNotFound, ChunkDataNotFound};

// 创建全局缓存实例，容量为2GB
pub static G_CHUNK_BLOCK_CACHE: LazyLock<SharedCache<String, Box<dyn CacheItem + Send + Sync>>> = LazyLock::new(|| new_shared_cache(2 * 1024 * 1024 * 1024));

/// Chunk 结构表示一个数据块
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Chunk {
    /// 内容的哈希
    pub hash: String,
    /// 块大小(字节)
    pub size: i32,
    /// 引用计数
    pub ref_count: i32,
    /// 所属BlockID
    pub block_id: String,
    /// 仅用于内存操作，不持久化
    #[serde(skip)]
    pub data: Vec<u8>,
}

impl Chunk {
    /// 创建新的数据块
    pub fn new(data: Vec<u8>) -> Self {
        let mut chunk = Chunk {
            hash: String::new(),
            size: data.len() as i32,
            ref_count: 0,
            block_id: "".to_string(),
            data,
        };
        chunk.hash = calc_hash(&chunk.data);
        chunk
    }
    /// 合并另一个块的数据到这个块
    pub fn merge(&mut self, other: &Chunk) {
        self.data.extend_from_slice(&other.data);
        self.size = self.data.len() as i32;
        self.hash = calc_hash(&self.data); // 重新计算哈希
    }
}

// 实现CacheItem特质，用于缓存和类型转换
impl CacheItem for Chunk {
    // 返回chunk的大小，用于缓存容量计算
    fn size(&self) -> usize {
        // 包括数据大小和其他字段的估计大小
        self.data.len() + 
        self.hash.len() + 
        self.block_id.len() + 
        12 // size(4) + ref_count(4) + 其他字段的估计(4)
    }
    
    // 转换为Any类型
    fn as_any(&self) -> &dyn Any {
        self
    }
    
    // 转换为Any类型的可变引用
    fn as_any_mut(&mut self) -> &mut dyn Any {
        self
    }
}

pub fn calc_hash( data: &[u8]) -> String {
    let hash = blake3::hash(data).to_hex().to_string();
    return hash; 
}

// 保留原有函数以保持兼容性
pub fn do_chunking(data: &[u8], fs: &DedupFS) -> Result<Vec<Chunk>> {
    info!("starting chunking process for data of size {} bytes", data.len());
    let config = &fs.chunk_conf;
    // 把 config 内容 打印到控制台
    info!("chunking configuration: {:?}", config);

    let mut chunks = Vec::new();
    
    if config.fixed_size {
        // 固定大小分块
        let chunk_size = config.avg_size as usize;
        for chunk_data in data.chunks(chunk_size) {
            let chunk = Chunk::new(chunk_data.to_vec());
            chunks.push(chunk);
        }
    } else {
        // 使用 FastCDC 进行内容定义分块
        let chunker = FastCDC::new(
            data,
            (config.min_size as usize).try_into().map_err(|e: std::num::TryFromIntError| {
                error!("chunk size conversion error for min_size {}: {}", config.min_size, e);
                crate::errors::new(crate::errors::ChunkSizeConversionError { size: config.min_size.to_string(), error: e.to_string() })
            })?,
            (config.avg_size as usize).try_into().map_err(|e: std::num::TryFromIntError| {
                error!("chunk size conversion error for avg_size {}: {}", config.avg_size, e);
                crate::errors::new(crate::errors::ChunkSizeConversionError { size: config.avg_size.to_string(), error: e.to_string() })
            })?,
            (config.max_size as usize).try_into().map_err(|e: std::num::TryFromIntError| {
                error!("chunk size conversion error for max_size {}: {}", config.max_size, e);
                crate::errors::new(crate::errors::ChunkSizeConversionError { size: config.max_size.to_string(), error: e.to_string() })
            })?,
        );
        
        for entry in chunker {
            let chunk_data = &data[entry.offset..entry.offset + entry.length];
            let chunk = Chunk::new(chunk_data.to_vec());
            chunks.push(chunk);
        }
    }

    // 后处理：合并太小的最后一个块
    if chunks.len() > 1 {
        let last_chunk = chunks.last().unwrap();
        let min_acceptable_size = (config.avg_size / 2) as i32;
        
        if last_chunk.size < min_acceptable_size {
            // 最后一个块太小，合并到前一个块
            info!("merging small last chunk (size: {}) with previous chunk", last_chunk.size);
            let last_chunk = chunks.pop().unwrap();
            let second_last_index = chunks.len() - 1;
            chunks[second_last_index].merge(&last_chunk);
        }
    }
    
    info!("chunking completed, generated {} chunks", chunks.len());
    Ok(chunks)
}

pub fn get_chunk_meta(hash: &str, fs: &DedupFS) -> Result<Chunk> {
    info!("retrieving chunk metadata with hash: {}", hash);
    
    // 将&str转换为String以便在缓存中使用
    let hash_string = hash.to_string();
    
    // 先从全局缓存中读取
    let cache_key = format!("{}:{}:{}", key_prefix::CHUNK, fs.id, hash_string);
    // 重新实现缓存读取逻辑，避免在trait object上调用clone
    if let Some(chunk_arc) = G_CHUNK_BLOCK_CACHE.get(&cache_key) {
        // 按照编译器建议，使用引用方式进行类型转换
        let chunk_box_ptr = &*chunk_arc as *const _ as *const Box<Chunk>;
        if !chunk_box_ptr.is_null() {
            // 安全地读取缓存中的Chunk数据
            let chunk_data = unsafe { &*chunk_box_ptr };
            let chunk = Chunk {
                hash: chunk_data.hash.clone(),
                size: chunk_data.size,
                ref_count: chunk_data.ref_count,
                block_id: chunk_data.block_id.clone(),
                data: chunk_data.data.clone(),
            };
            info!("chunk metadata found in cache with hash: {}, size: {}, ref_count: {}, block_id: {}", 
                 hash, chunk.size, chunk.ref_count, chunk.block_id);
            return Ok(chunk);
        }
    }
    
    // 缓存未命中，从KVStore获取chunk元数据
    info!("retrieving chunk metadata from kvstore with hash: {}", hash);
    let chunk_key = make_prefixed_key(key_prefix::CHUNK, hash.as_bytes());
    
    let chunk = fs.kv_store.get::<Chunk>(&chunk_key)?
        .ok_or_else(|| {
            error!("chunk metadata not found with hash: {}", hash);
            ChunkNotFound { hash: hash.to_string() }
        })?;
    
    // 使用全局缓存，key为fs.id + ":" + hash
    let cache_key = format!("{}:{}:{}",key_prefix::CHUNK, fs.id, hash_string);
    G_CHUNK_BLOCK_CACHE.put(cache_key, Box::new(chunk.clone()), |item| item.size());
    
    info!("successfully retrieved chunk metadata with hash: {}, size: {}, ref_count: {}, block_id: {}", 
         hash, chunk.size, chunk.ref_count, chunk.block_id);
    
    Ok(chunk)
}

pub fn get_chunk_data(hash: &str, fs: &DedupFS) -> Result<Chunk> {
    info!("retrieving chunk with hash: {}", hash);
    
    // 首先调用get_chunk_meta获取元数据
    let mut chunk = get_chunk_meta(hash, fs)?;
    
    // 检查chunk.data是否已经有数据
    if !chunk.data.is_empty() {
        info!("chunk data already available for hash: {}, size: {} bytes", hash, chunk.data.len());
        return Ok(chunk);
    }
    
    info!("chunk data is empty for hash: {}, proceeding to load from block", hash);
    
    // 从元数据中获取block id（克隆以避免部分移动）
    let block_id = chunk.block_id.clone();
    error!("found chunk metadata, retrieving from block: {}", block_id);
    
    // 调用read_block函数获取block数据
    let block = read_block(&block_id, fs)?;
    
    // 从block中找到对应的chunk数据
    // 计算chunk在block中的偏移量
    let mut offset = 0;
    let mut target_chunk_data = None;
    
    for chunk_info in &block.header.chunk_list {
        if chunk_info.hash == hash {
            // 找到目标chunk，从block数据中提取
            let chunk_size = chunk_info.size as usize;
            info!("found chunk in block, size: {}, offset: {}", chunk_size, offset);
            if offset + chunk_size <= block.data.len() {
                target_chunk_data = Some(block.data[offset..offset + chunk_size].to_vec());
            } else {
                error!("chunk size {} exceeds available block data len {} at offset {}", chunk_size, block.data.len(), offset);
            }
            break;
        }
        // 更新偏移量
        offset += chunk_info.size as usize;
    }
    
    // 检查是否找到chunk数据
    let chunk_data = target_chunk_data
        .ok_or_else(|| {
            error!("chunk data not found in block {} for hash: {}", block_id, hash);
            ChunkDataNotFound { block_id, hash: hash.to_string() }
        })?;
    
    // 设置chunk数据
    chunk.data = chunk_data;
    
    // 使用全局缓存，key为fs.id + ":" + hash
    let cache_key = format!("{}:{}:{}",key_prefix::CHUNK, fs.id, hash);
    G_CHUNK_BLOCK_CACHE.put(cache_key, Box::new(chunk.clone()), |item| item.size());
    
    info!("successfully retrieved chunk with hash: {}, size: {}", hash, chunk.size);
    
    Ok(chunk)
}

pub fn remove_chunk(hash: &str, fs: &DedupFS) -> Result<()> {
    info!("remove_chunk - removing chunk with hash: {}", hash);
    
    // 先从kv_store 中 获取 chunk
    let chunk_key = make_prefixed_key(key_prefix::CHUNK, hash.as_bytes());
    let mut chunk = fs.kv_store.get::<Chunk>(&chunk_key)?
        .ok_or_else(|| {
            error!("remove_chunk - chunk not found with hash: {}", hash);
            ChunkNotFound { hash: hash.to_string() }
        })?;
    
    // 对引用计数 ref_count - 1
    chunk.ref_count -= 1;
    info!("remove_chunk - chunk {} ref_count updated to {}", hash, chunk.ref_count);
    
    if chunk.ref_count == 0 {
        // 如果引用计数为0，需要调用 block:remove_chunk 从对应的 block 中删除该chunk
        info!("remove_chunk - chunk {} ref_count is 0, removing from block {}", hash, chunk.block_id);
        
        // 从block中删除chunk数据
        crate::block::remove_chunk_from_block(&chunk, fs)?;
        
        // 从kv_store中删除chunk元数据
        fs.kv_store.del(&chunk_key)?;
        info!("remove_chunk - chunk {} deleted from kv_store", hash);
        
        // 从缓存中移除
        let cache_key = format!("{}:{}:{}", key_prefix::CHUNK, fs.id, hash);
        G_CHUNK_BLOCK_CACHE.del(&cache_key);
        info!("remove_chunk - chunk {} removed from cache", hash);
    } else {
        // 如果引用计数大于0，保存更新后的chunk到kv_store
        fs.kv_store.set(&chunk_key, &chunk)?;
        info!("remove_chunk - updated chunk {} with ref_count {} saved to kv_store", hash, chunk.ref_count);
        
        // 更新缓存
        let cache_key = format!("{}:{}:{}",key_prefix::CHUNK, fs.id, hash);
        G_CHUNK_BLOCK_CACHE.put(cache_key, Box::new(chunk.clone()), |item| item.size());
    }
    
    info!("remove_chunk - successfully processed chunk {}", hash);
    Ok(())
}

File: dfs/chunk.go
package dfs

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"

	fastcdc "github.com/PlakarKorp/go-cdc-chunkers"
	_ "github.com/PlakarKorp/go-cdc-chunkers/chunkers/fastcdc"
	"github.com/cespare/xxhash/v2"

	"github.com/mageg-x/dedupfs/internal/cache"
)

var (
	G_CHUNK_CACHE = cache.NewCache[string, *Chunk](1024*1024*1024, true)
)

// Chunk 表示一个数据块
type Chunk struct {
	Hash     string `json:"hash"`
	Size     int32  `json:"size"`
	RefCount int32  `json:"ref_count"`
	BlockID  string `json:"block_id"`
	Data     []byte `json:"-"` // 仅内存使用，不持久化
}

// MarshalJSONWithData 返回包含 Data 字段的 JSON
func (c *Chunk) MarshalJSONWithData() ([]byte, error) {
	type Alias Chunk
	return json.Marshal(&struct {
		Data []byte `json:"data"`
		*Alias
	}{
		Data:  c.Data,
		Alias: (*Alias)(c),
	})
}
func (c *Chunk) UnmarshalJSONWithData(data []byte) error {
	type Alias Chunk
	aux := &struct {
		Data []byte `json:"data"`
		*Alias
	}{
		Alias: (*Alias)(c),
	}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	c.Data = aux.Data
	return nil
}

// 实现 CacheItem
func (c *Chunk) Len() int {
	return len(c.Data) + len(c.Hash) + len(c.BlockID) + 12 // 估算其他字段
}

// Merge 合并另一个 chunk 的数据
func (c *Chunk) Merge(other *Chunk) {
	logger.Debugf("merging chunks: %s + %s", c.Hash, other.Hash)
	c.Data = append(c.Data, other.Data...)
	c.Size = int32(len(c.Data))
	c.Hash = calcHash(c.Data)
	logger.Debugf("merged chunk hash: %s, size: %d", c.Hash, c.Size)
}

// calcHash 使用 blake3
func calcHash(data []byte) string {
	// fp := blake3.Sum256(data)
	// hash := hex.EncodeToString(fp[:20])
	fp := xxhash.Sum64(data)
	hash := fmt.Sprintf("%x", fp)
	return hash
}

// NewChunk 创建新块
func NewChunk(data []byte) *Chunk {
	hash := calcHash(data)
	logger.Debugf("creating new chunk with hash %s, size: %d bytes", hash, len(data))
	return &Chunk{
		Hash:     hash,
		Size:     int32(len(data)),
		RefCount: 0,
		BlockID:  "",
		Data:     data,
	}
}

// GetChunkMeta 获取元数据（不含 Data）
func GetChunkMeta(hash string, fs *DedupFS) (*Chunk, error) {
	logger.Debugf("getting chunk meta for hash %s in filesystem %s", hash, fs.ID)
	cacheKey := fmt.Sprintf("chunk:%s:%s", fs.ID, hash)
	if chunk, ok := G_CHUNK_CACHE.Get(cacheKey); ok && chunk != nil {
		logger.Debugf("chunk meta %s found in cache", hash)
		return chunk, nil
	}

	// 从 KVStore 读取
	chunkKey := fmt.Sprintf("chunk:%s", hash)
	var chunk Chunk
	if err := fs.KVStore.Get(chunkKey, &chunk); err != nil {
		logger.Warnf("chunk %s not found in kvstore", hash)
		return nil, fmt.Errorf("chunk not found")
	}

	G_CHUNK_CACHE.Put(cacheKey, &chunk)
	logger.Debugf("loaded chunk meta %s from kvstore", hash)

	return &chunk, nil
}

// GetChunkData 获取完整 chunk（含 Data）
func GetChunkData(hash string, fs *DedupFS) (*Chunk, error) {
	logger.Debugf("getting chunk data for hash %s in filesystem %s", hash, fs.ID)
	if len(hash) == 0 || fs == nil {
		logger.Errorf("invalid arguments: hash=%s, fs=%v", hash, fs)
		return nil, fmt.Errorf("invalid arguments")
	}
	// 先取元数据
	chunk, err := GetChunkMeta(hash, fs)
	if err != nil || chunk == nil {
		logger.Errorf("failed to get chunk meta for %s: %v", hash, err)
		return nil, err
	}

	if len(chunk.Data) > 0 {
		logger.Debugf("chunk %s data already available", hash)
		return chunk, nil
	}

	// 读 block
	block, err := ReadBlock(chunk.BlockID, fs)
	if err != nil || block == nil {
		logger.Errorf("failed to read block %s for chunk %s: %v", chunk.BlockID, hash, err)
		return nil, err
	}

	// 在 block 中找 chunk
	offset := 0
	var chunkData []byte
	for _, c := range block.Header.ChunkList {
		if c.Hash == hash {
			end := offset + int(c.Size)
			if end > len(block.Data) {
				logger.Errorf("chunk %s out of bounds in block %s", hash, chunk.BlockID)
				return nil, fmt.Errorf("chunk out of bounds in block %s", chunk.BlockID)
			}
			chunkData = block.Data[offset:end]
			break
		}
		offset += int(c.Size)
	}

	if chunkData == nil {
		logger.Errorf("failed to read chunk data for %s from block %s", hash, chunk.BlockID)
		return nil, fmt.Errorf("failed to read chunk data")
	}

	chunk.Data = chunkData
	cacheKey := fmt.Sprintf("chunk:%s:%s", fs.ID, hash)
	G_CHUNK_CACHE.Put(cacheKey, chunk)
	logger.Debugf("successfully loaded chunk data for %s, size: %d bytes", hash, len(chunkData))

	return chunk, nil
}

// RemoveChunk 减引用计数，必要时删除
func RemoveChunk(hash string, fs *DedupFS) error {
	logger.Debugf("removing chunk %s from filesystem %s", hash, fs.ID)
	chunkKey := fmt.Sprintf("chunk:%s", hash)
	var chunk Chunk
	if err := fs.KVStore.Get(chunkKey, &chunk); err != nil {
		logger.Errorf("chunk %s not found for removal: %v", hash, err)
		return fmt.Errorf("chunk not found")
	}

	chunk.RefCount--
	logger.Debugf("chunk %s ref count decremented to %d", hash, chunk.RefCount)
	if chunk.RefCount <= 0 {
		logger.Debugf("chunk %s ref count zero, removing completely", hash)
		// 从 block 中删除
		if err := RemoveChunkFromBlock(&chunk, fs); err != nil {
			logger.Errorf("failed to remove chunk %s from block: %v", hash, err)
			return err
		}
		// 删除元数据
		if err := fs.KVStore.Del(chunkKey); err != nil {
			logger.Errorf("failed to delete chunk %s metadata: %v", hash, err)
			return err
		}
		// 清缓存
		cacheKey := fmt.Sprintf("chunk:%s:%s", fs.ID, hash)
		G_CHUNK_CACHE.Del(cacheKey)
		logger.Infof("chunk %s completely removed", hash)
	} else {
		// 更新引用计数
		if err := fs.KVStore.Set(chunkKey, &chunk); err != nil {
			logger.Errorf("failed to update chunk %s ref count: %v", hash, err)
			return err
		}
		cacheKey := fmt.Sprintf("chunk:%s:%s", fs.ID, hash)
		G_CHUNK_CACHE.Put(cacheKey, &chunk)
		logger.Debugf("updated chunk %s ref count to %d", hash, chunk.RefCount)
	}

	return nil
}

func PutChunks(chunks []*Chunk, fs *DedupFS) error {
	logger.Debugf("putting %d chunks into block", len(chunks))
	for _, chunk := range chunks {
		if int32(len(chunk.Data)) != chunk.Size {
			logger.Errorf("chunk size mismatch: %s expected=%d actual=%d", chunk.Hash, chunk.Size, len(chunk.Data))
			return fmt.Errorf("hash=%s expected=%d actual=%d", chunk.Hash, chunk.Size, len(chunk.Data))
		}

		// 从 KVStore 读取
		chunkKey := fmt.Sprintf("chunk:%s", chunk.Hash)
		var c Chunk
		var exist bool
		if err := fs.KVStore.Get(chunkKey, &c); err != nil {
			logger.Warnf("chunk %s not found in kvstore", chunk.Hash)
			chunk.RefCount = 1
		} else {
			chunk.RefCount = c.RefCount + 1
			chunk.BlockID = c.BlockID
			exist = true
			logger.Debugf("chunk %s:%d:%d found in kvstore", chunk.Hash, chunk.RefCount, chunk.Size)
		}

		if !exist {
			logger.Debugf("chunk %s size %d adding to block", chunk.Hash, len(chunk.Data))
			if err := AddChunkToBlock(chunk, fs); err != nil {
				logger.Errorf("failed to add chunk %s to block: %v", chunk.Hash, err)
				return fmt.Errorf("failed to add chunk %s to block: %w", chunk.Hash, err)
			}
		}

		logger.Debugf("chunk %s:%d:%d block %s saved to kvstore", chunk.Hash, chunk.RefCount, chunk.Size, chunk.BlockID)

		if err := fs.KVStore.Set(chunkKey, chunk); err != nil {
			logger.Errorf("failed to save chunk %s metadata: %v", chunk.Hash, err)
			return fmt.Errorf("failed to save chunk %s metadata: %w", chunk.Hash, err)
		}

		cacheKey := fmt.Sprintf("chunk:%s:%s", fs.ID, chunk.Hash)
		G_CHUNK_CACHE.Put(cacheKey, chunk)
	}
	return nil
}

// DoChunking 分块主函数
func DoChunking(data []byte, fs *DedupFS) ([]*Chunk, error) {
	logger.Debugf("starting chunking process for %d bytes of data", len(data))
	cfg := fs.ChunkConf
	var chunks []*Chunk

	if cfg.FixedSize {
		logger.Debugf("using fixed size chunking with size %d", cfg.AvgSize)
		chunkSize := int(cfg.AvgSize)
		for i := 0; i < len(data); i += chunkSize {
			end := i + chunkSize
			if end > len(data) {
				end = len(data)
			}
			chunk := NewChunk(data[i:end])
			chunks = append(chunks, chunk)
		}
	} else {
		logger.Infof("using cdc chunking: min=%d, avg=%d, max=%d", cfg.MinSize, cfg.AvgSize, cfg.MaxSize)
		chunker, err := fastcdc.NewChunker("fastcdc", bytes.NewReader(data), &fastcdc.ChunkerOpts{
			MinSize:    int(cfg.MinSize),
			NormalSize: int(cfg.AvgSize),
			MaxSize:    int(cfg.MaxSize),
		})

		if err != nil || chunker == nil {
			logger.Errorf("fastcdc init failed: %v", err)
			return nil, fmt.Errorf("fastcdc init failed: %w", err)
		}

		for {
			chunkData, err := chunker.Next()
			if err != nil && err != io.EOF {
				logger.Errorf("cdc chunk error: %v", err)
				return nil, fmt.Errorf("cdc chunk error: %w", err)
			}
			if len(chunkData) > 0 {
				chunk := NewChunk(chunkData)
				chunks = append(chunks, chunk)
			}

			if err == io.EOF {
				break
			}
		}
	}

	// 合并过小的最后一个块
	if len(chunks) > 1 {
		last := chunks[len(chunks)-1]
		minAcceptable := cfg.AvgSize / 2
		if int64(last.Size) < minAcceptable {
			logger.Debugf("last chunk too small (%d bytes), merging with previous chunk", last.Size)
			prev := chunks[len(chunks)-2]
			prev.Merge(last)
			chunks = chunks[:len(chunks)-1]
		}
	}

	logger.Debugf("chunking completed, created %d chunks", len(chunks))
	return chunks, nil
}
